### PR TITLE
PIM-9711 check that a category root isn't linked to a user or a channel before moving it to a sub-category

### PR DESCRIPTION
**Pull request by pchasle**
Update the `CategoryUpdater` to check that a category is not linked to a `User` or a `Channel` before moving it to be a sub-category.
It perform the check only when the category is a root one to avoid unnecessary checks on sub-categories that should NOT be linked to user or channel.

### CODE REVIEW

File: tests/back/Channel/Integration/Channel/Query/Sql/SqlIsCategoryTreeLinkedToChannelIntegration.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    private SqlIsCategoryTreeLinkedToChannel $isCategoryTreeLinkedToChannel;
+
+    public function setUp(): void
+    {
+        parent::setUp();
+
+        $this->isCategoryTreeLinkedToChannel = $this->get('pim_channel.query.sql.is_category_tree_linked_to_channel');
+    }
+
+    public function test_it_counts_the_number_channels_linked_to_a_category(): void
```
Comment (mmetayer) [inline tests/back/Channel/Integration/Channel/Query/Sql/SqlIsCategoryTreeLinkedToChannelIntegration.php:27]:
```suggestion
    public function test_it_checks_if_a_category_is_linked_to_a_channel(): void
```

File: tests/back/Channel/Integration/Channel/Query/Sql/SqlIsCategoryTreeLinkedToChannelIntegration.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    {
+        $category = $this->createCategory(['code' => 'clothes']);
+        $this->createChannel(['code' => 'mobile', 'category_tree' => 'clothes']);
+
+        $isLinked = $this->isCategoryTreeLinkedToChannel->byCategoryTreeId($category->getId());
+
+        $this->assertTrue($isLinked);
```
Comment (mmetayer) [inline tests/back/Channel/Integration/Channel/Query/Sql/SqlIsCategoryTreeLinkedToChannelIntegration.php:34]: could you also test the opposite please? e.g `Assert::assertFalse($this->isCategoryTreeLinkedToChannel->byCategoryTreeId($anotherCategory->getId()));`

File: src/Akeneo/UserManagement/Bundle/Storage/Query/SqlIsCategoryTreeLinkedToUser.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Akeneo\UserManagement\Bundle\Storage\Query;
```
Comment (mmetayer) [inline src/Akeneo/UserManagement/Bundle/Storage/Query/SqlIsCategoryTreeLinkedToUser.php:5]: I would have put it in the `Akeneo\Channel`  namespace as well, but I guess `Akeneo\UserManagement`  is ok too

File: tests/back/UserManagement/Integration/Bundle/Query/SqlIsCategoryTreeLinkedToUserIntegration.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        $this->createUser([
+            'username' => 'julia',
+            'email' => 'julia@akeneo.com',
+            'password' => 'a_password',
+            'default_category_tree' => 'clothes'
+        ]);
+
+        $isLinked = $this->isCategoryTreeLinkedToUser->byCategoryTreeId($category->getId());
+
+        $this->assertTrue($isLinked);
```
Comment (mmetayer) [inline tests/back/UserManagement/Integration/Bundle/Query/SqlIsCategoryTreeLinkedToUserIntegration.php:39]: same: could you add an assertion for the opposite?

File: tests/back/Channel/Integration/Channel/Query/Sql/SqlIsCategoryTreeLinkedToChannelIntegration.php [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    protected function getConfiguration(): Configuration
+    {
+        return $this->catalog->useMinimalCatalog();
+    }
+
+    private function createChannel(array $data): ChannelInterface
+    {
+        $channel = $this->get('pim_catalog.factory.channel')->create();
+        $this->get('pim_catalog.updater.channel')->update($channel, $data);
+        $this->get('validator')->validate($channel);
```
Comment (mmetayer) [inline tests/back/Channel/Integration/Channel/Query/Sql/SqlIsCategoryTreeLinkedToChannelIntegration.php:46]: this won't interrupt the workflow, the channel will be saved even if it's not valid :)

File: tests/back/UserManagement/Integration/Bundle/Query/SqlIsCategoryTreeLinkedToUserIntegration.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    protected function getConfiguration(): Configuration
+    {
+        return $this->catalog->useMinimalCatalog();
+    }
+
+    private function createUser(array $data): UserInterface
+    {
+        $user = $this->get('pim_user.factory.user')->create();
+        $this->get('pim_user.updater.user')->update($user, $data);
+        $this->get('validator')->validate($user);
```
Comment (mmetayer) [inline tests/back/UserManagement/Integration/Bundle/Query/SqlIsCategoryTreeLinkedToUserIntegration.php:51]: same remark as for the other integration test

**Review by mmetayer — APPROVED** [review]

File: tests/back/Channel/Integration/Channel/Query/Sql/SqlIsCategoryTreeLinkedToChannelIntegration.php (hunk 5 above)
Comment (pchasle) [reply tests/back/Channel/Integration/Channel/Query/Sql/SqlIsCategoryTreeLinkedToChannelIntegration.php:46]: Ah yes, you are right :facepalm:

**Review by jmleroux — APPROVED** [review]

**Pull request merged**